Handle connection error "Cannot connect to host gateway.autonolas"

Currently, this connection error produces: `ValueError: The mech's tools have not been set.`

I also observed this. The non-availability of the gateway should not affect the agents ability to run. Of course, it will affect its behaviour.

Full trace:

```
[2023-11-15 14:34:29,515][ERROR] [agent] Response: "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/connector.py\", line 969, in _wrap_create_connection\n    return await self._loop.create_connection(*args, **kwargs)  # type: ignore  # noqa\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/asyncio/base_events.py\", line 1085, in create_connection\n    raise exceptions[0]\n  File \"/usr/local/lib/python3.11/asyncio/base_events.py\", line 1069, in create_connection\n    sock = await self._connect_sock(\n           ^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/asyncio/base_events.py\", line 973, in _connect_sock\n    await self.sock_connect(sock, address)\n  File \"/usr/local/lib/python3.11/asyncio/selector_events.py\", line 634, in sock_connect\n    return await fut\n           ^^^^^^^^^\n  File \"/usr/local/lib/python3.11/asyncio/selector_events.py\", line 674, in _sock_connect_cb\n    raise OSError(err, f'Connect call failed {address}')\nConnectionRefusedError: [Errno 111] Connect call failed ('3.11.123.203', 443)\n\nThe above exception was the direct cause of the following exception:\n\nTraceback (most recent call last):\n  File \"/root/agent/vendor/valory/connections/http_client/connection.py\", line 190, in _http_request_task\n    resp = await asyncio.wait_for(\n           ^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/asyncio/tasks.py\", line 489, in wait_for\n    return fut.result()\n           ^^^^^^^^^^^^\n  File \"/root/agent/vendor/valory/connections/http_client/connection.py\", line 235, in _perform_http_request\n    async with session.request(\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/client.py\", line 1117, in __aenter__\n    self._resp = await self._coro\n                 ^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/client.py\", line 520, in _request\n    conn = await self._connector.connect(\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/connector.py\", line 535, in connect\n    proto = await self._create_connection(req, traces, timeout)\n            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/connector.py\", line 892, in _create_connection\n    _, proto = await self._create_direct_connection(req, traces, timeout)\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/connector.py\", line 1051, in _create_direct_connection\n    raise last_exc\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/connector.py\", line 1020, in _create_direct_connection\n    transp, proto = await self._wrap_create_connection(\n                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/connector.py\", line 975, in _wrap_create_connection\n    raise client_error(req.connection_key, exc) from exc\naiohttp.client_exceptions.ClientConnectorError: Cannot connect to host gateway.autonolas.tech:443 ssl:<ssl.SSLContext object at 0x7f11f86dc830> [Connect call failed ('3.11.123.203', 443)]\n"
[2023-11-15 14:34:29,515] [ERROR] [agent] Response: "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/connector.py\", line 969, in _wrap_create_connection\n    return await self._loop.create_connection(*args, **kwargs)  # type: ignore  # noqa\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/asyncio/base_events.py\", line 1085, in create_connection\n    raise exceptions[0]\n  File \"/usr/local/lib/python3.11/asyncio/base_events.py\", line 1069, in create_connection\n    sock = await self._connect_sock(\n           ^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/asyncio/base_events.py\", line 973, in _connect_sock\n    await self.sock_connect(sock, address)\n  File \"/usr/local/lib/python3.11/asyncio/selector_events.py\", line 634, in sock_connect\n    return await fut\n           ^^^^^^^^^\n  File \"/usr/local/lib/python3.11/asyncio/selector_events.py\", line 674, in _sock_connect_cb\n    raise OSError(err, f'Connect call failed {address}')\nConnectionRefusedError: [Errno 111] Connect call failed ('3.11.123.203', 443)\n\nThe above exception was the direct cause of the following exception:\n\nTraceback (most recent call last):\n  File \"/root/agent/vendor/valory/connections/http_client/connection.py\", line 190, in _http_request_task\n    resp = await asyncio.wait_for(\n           ^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/asyncio/tasks.py\", line 489, in wait_for\n    return fut.result()\n           ^^^^^^^^^^^^\n  File \"/root/agent/vendor/valory/connections/http_client/connection.py\", line 235, in _perform_http_request\n    async with session.request(\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/client.py\", line 1117, in __aenter__\n    self._resp = await self._coro\n                 ^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/client.py\", line 520, in _request\n    conn = await self._connector.connect(\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/connector.py\", line 535, in connect\n    proto = await self._create_connection(req, traces, timeout)\n            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/connector.py\", line 892, in _create_connection\n    _, proto = await self._create_direct_connection(req, traces, timeout)\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/connector.py\", line 1051, in _create_direct_connection\n    raise last_exc\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/connector.py\", line 1020, in _create_direct_connection\n    transp, proto = await self._wrap_create_connection(\n                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/usr/local/lib/python3.11/site-packages/aiohttp/connector.py\", line 975, in _wrap_create_connection\n    raise client_error(req.connection_key, exc) from exc\naiohttp.client_exceptions.ClientConnectorError: Cannot connect to host gateway.autonolas.tech:443 ssl:<ssl.SSLContext object at 0x7f11f86dc830> [Connect call failed ('3.11.123.203', 443)]\n"
[2023-11-15 14:34:29,529][ERROR] [agent] Retries were exceeded while trying to get the mech agent's data.
[2023-11-15 14:34:29,529] [ERROR] [agent] Retries were exceeded while trying to get the mech agent's data.
[2023-11-15 14:34:29,557][ERROR] [agent] <An error occurred during act of behaviour valory/trader_abci:0.1.0/TraderConsensusBehaviour:
Traceback (most recent call last):

  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")

ValueError: The mech's tools have not been set.


During handling of the above exception, another exception occurred:


Traceback (most recent call last):

  File "/usr/local/lib/python3.11/site-packages/aea/skills/base.py", line 440, in act_wrapper
    raise AEAActException(

aea.exceptions.AEAActException: An error occurred during act of behaviour valory/trader_abci:0.1.0/ToolSelectionBehaviour:
Traceback (most recent call last):

  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")

ValueError: The mech's tools have not been set.

> raised during `<bound method Behaviour.act_wrapper of <behaviours.TraderConsensusBehaviour object at 0x7f11f817a190>>`
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aea/skills/base.py", line 435, in act_wrapper
    self.act()
  File "/root/agent/vendor/valory/skills/abstract_round_abci/behaviour_utils.py", line 318, in act
    self.__handle_waiting_for_message()
  File "/root/agent/vendor/valory/skills/abstract_round_abci/behaviour_utils.py", line 354, in __handle_waiting_for_message
    self.__get_generator_act().send(self.__message)
  File "/root/agent/vendor/valory/skills/abstract_round_abci/behaviour_utils.py", line 731, in async_act_wrapper
    yield from self.async_act()
  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 277, in async_act
    selected_tool = yield from self._select_tool()
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 255, in _select_tool
    self._set_policy()
  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 204, in _set_policy
    self._adjust_policy_tools(local_tools)
  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 184, in _adjust_policy_tools
    removed_idx = [idx for idx in reversed_idx if local[idx] not in self.mech_tools]
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 184, in <listcomp>
    removed_idx = [idx for idx in reversed_idx if local[idx] not in self.mech_tools]
                                                                    ^^^^^^^^^^^^^^^
  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")
ValueError: The mech's tools have not been set.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aea/skills/base.py", line 435, in act_wrapper
    self.act()
  File "/root/agent/vendor/valory/skills/abstract_round_abci/behaviours.py", line 368, in act
    self.current_behaviour.act_wrapper()
  File "/usr/local/lib/python3.11/site-packages/aea/skills/base.py", line 440, in act_wrapper
    raise AEAActException(
aea.exceptions.AEAActException: An error occurred during act of behaviour valory/trader_abci:0.1.0/ToolSelectionBehaviour:
Traceback (most recent call last):

  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")

ValueError: The mech's tools have not been set.


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aea/agent_loop.py", line 286, in _execution_control
    return fn(*(args or []), **(kwargs or {}))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aea/skills/base.py", line 440, in act_wrapper
    raise AEAActException(
aea.exceptions.AEAActException: An error occurred during act of behaviour valory/trader_abci:0.1.0/TraderConsensusBehaviour:
Traceback (most recent call last):

  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")

ValueError: The mech's tools have not been set.


During handling of the above exception, another exception occurred:


Traceback (most recent call last):

  File "/usr/local/lib/python3.11/site-packages/aea/skills/base.py", line 440, in act_wrapper
    raise AEAActException(

aea.exceptions.AEAActException: An error occurred during act of behaviour valory/trader_abci:0.1.0/ToolSelectionBehaviour:
Traceback (most recent call last):

  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")

ValueError: The mech's tools have not been set.


[2023-11-15 14:34:29,557] [ERROR] [agent] <An error occurred during act of behaviour valory/trader_abci:0.1.0/TraderConsensusBehaviour:
Traceback (most recent call last):

  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")

ValueError: The mech's tools have not been set.


During handling of the above exception, another exception occurred:


Traceback (most recent call last):

  File "/usr/local/lib/python3.11/site-packages/aea/skills/base.py", line 440, in act_wrapper
    raise AEAActException(

aea.exceptions.AEAActException: An error occurred during act of behaviour valory/trader_abci:0.1.0/ToolSelectionBehaviour:
Traceback (most recent call last):

  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")

ValueError: The mech's tools have not been set.

> raised during `<bound method Behaviour.act_wrapper of <behaviours.TraderConsensusBehaviour object at 0x7f11f817a190>>`
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aea/skills/base.py", line 435, in act_wrapper
    self.act()
  File "/root/agent/vendor/valory/skills/abstract_round_abci/behaviour_utils.py", line 318, in act
    self.__handle_waiting_for_message()
  File "/root/agent/vendor/valory/skills/abstract_round_abci/behaviour_utils.py", line 354, in __handle_waiting_for_message
    self.__get_generator_act().send(self.__message)
  File "/root/agent/vendor/valory/skills/abstract_round_abci/behaviour_utils.py", line 731, in async_act_wrapper
    yield from self.async_act()
  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 277, in async_act
    selected_tool = yield from self._select_tool()
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 255, in _select_tool
    self._set_policy()
  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 204, in _set_policy
    self._adjust_policy_tools(local_tools)
  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 184, in _adjust_policy_tools
    removed_idx = [idx for idx in reversed_idx if local[idx] not in self.mech_tools]
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 184, in <listcomp>
    removed_idx = [idx for idx in reversed_idx if local[idx] not in self.mech_tools]
                                                                    ^^^^^^^^^^^^^^^
  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")
ValueError: The mech's tools have not been set.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aea/skills/base.py", line 435, in act_wrapper
    self.act()
  File "/root/agent/vendor/valory/skills/abstract_round_abci/behaviours.py", line 368, in act
    self.current_behaviour.act_wrapper()
  File "/usr/local/lib/python3.11/site-packages/aea/skills/base.py", line 440, in act_wrapper
    raise AEAActException(
aea.exceptions.AEAActException: An error occurred during act of behaviour valory/trader_abci:0.1.0/ToolSelectionBehaviour:
Traceback (most recent call last):

  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")

ValueError: The mech's tools have not been set.


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aea/agent_loop.py", line 286, in _execution_control
    return fn(*(args or []), **(kwargs or {}))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aea/skills/base.py", line 440, in act_wrapper
    raise AEAActException(
aea.exceptions.AEAActException: An error occurred during act of behaviour valory/trader_abci:0.1.0/TraderConsensusBehaviour:
Traceback (most recent call last):

  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")

ValueError: The mech's tools have not been set.


During handling of the above exception, another exception occurred:


Traceback (most recent call last):

  File "/usr/local/lib/python3.11/site-packages/aea/skills/base.py", line 440, in act_wrapper
    raise AEAActException(

aea.exceptions.AEAActException: An error occurred during act of behaviour valory/trader_abci:0.1.0/ToolSelectionBehaviour:
Traceback (most recent call last):

  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")

ValueError: The mech's tools have not been set.


[2023-11-15 14:34:29,627] [INFO] [agent] HTTP Client has shutdown on port: 8000.
[2023-11-15 14:34:29,627][INFO] [agent] HTTP Client has shutdown on port: 8000.
[2023-11-15 14:34:29,633] [ERROR] [agent] exception on node client close
Traceback (most recent call last):
  File "/root/agent/vendor/valory/connections/p2p_libp2p_client/connection.py", line 501, in disconnect
    await self._node_client.close()
  File "/root/agent/vendor/valory/connections/p2p_libp2p_client/connection.py", line 268, in close
    await self.pipe.close()
  File "/usr/local/lib/python3.11/site-packages/aea/helpers/pipe.py", line 600, in close
    await self._sock.close()
  File "/usr/local/lib/python3.11/site-packages/aea/helpers/pipe.py", line 325, in close
    await self._writer.drain()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 377, in drain
    await self._protocol._drain_helper()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 166, in _drain_helper
    raise ConnectionResetError('Connection lost')
ConnectionResetError: Connection lost
[2023-11-15 14:34:29,633][ERROR] [agent] exception on node client close
Traceback (most recent call last):
  File "/root/agent/vendor/valory/connections/p2p_libp2p_client/connection.py", line 501, in disconnect
    await self._node_client.close()
  File "/root/agent/vendor/valory/connections/p2p_libp2p_client/connection.py", line 268, in close
    await self.pipe.close()
  File "/usr/local/lib/python3.11/site-packages/aea/helpers/pipe.py", line 600, in close
    await self._sock.close()
  File "/usr/local/lib/python3.11/site-packages/aea/helpers/pipe.py", line 325, in close
    await self._writer.drain()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 377, in drain
    await self._protocol._drain_helper()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 166, in _drain_helper
    raise ConnectionResetError('Connection lost')
ConnectionResetError: Connection lost
[2023-11-15 14:34:29,651][ERROR] [agent] Error while disconnecting <class 'connection_module.P2PLibp2pClientConnection'>: Connection lost
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aea/multiplexer.py", line 454, in _disconnect_all
    await asyncio.wait_for(
  File "/usr/local/lib/python3.11/asyncio/tasks.py", line 489, in wait_for
    return fut.result()
           ^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aea/multiplexer.py", line 481, in _disconnect_one
    await connection.disconnect()
  File "/root/agent/vendor/valory/connections/p2p_libp2p_client/connection.py", line 501, in disconnect
    await self._node_client.close()
  File "/root/agent/vendor/valory/connections/p2p_libp2p_client/connection.py", line 268, in close
    await self.pipe.close()
  File "/usr/local/lib/python3.11/site-packages/aea/helpers/pipe.py", line 600, in close
    await self._sock.close()
  File "/usr/local/lib/python3.11/site-packages/aea/helpers/pipe.py", line 325, in close
    await self._writer.drain()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 377, in drain
    await self._protocol._drain_helper()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 166, in _drain_helper
    raise ConnectionResetError('Connection lost')
ConnectionResetError: Connection lost
[2023-11-15 14:34:29,651] [ERROR] [agent] Error while disconnecting <class 'connection_module.P2PLibp2pClientConnection'>: Connection lost
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/aea/multiplexer.py", line 454, in _disconnect_all
    await asyncio.wait_for(
  File "/usr/local/lib/python3.11/asyncio/tasks.py", line 489, in wait_for
    return fut.result()
           ^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/aea/multiplexer.py", line 481, in _disconnect_one
    await connection.disconnect()
  File "/root/agent/vendor/valory/connections/p2p_libp2p_client/connection.py", line 501, in disconnect
    await self._node_client.close()
  File "/root/agent/vendor/valory/connections/p2p_libp2p_client/connection.py", line 268, in close
    await self.pipe.close()
  File "/usr/local/lib/python3.11/site-packages/aea/helpers/pipe.py", line 600, in close
    await self._sock.close()
  File "/usr/local/lib/python3.11/site-packages/aea/helpers/pipe.py", line 325, in close
    await self._writer.drain()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 377, in drain
    await self._protocol._drain_helper()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 166, in _drain_helper
    raise ConnectionResetError('Connection lost')
ConnectionResetError: Connection lost
Stopping AEA 'agent' ...
AEA 'agent' stopped.
Error: AEA was terminated cause exception `An error occurred during act of behaviour valory/trader_abci:0.1.0/TraderConsensusBehaviour:
Traceback (most recent call last):

  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")

ValueError: The mech's tools have not been set.


During handling of the above exception, another exception occurred:


Traceback (most recent call last):

  File "/usr/local/lib/python3.11/site-packages/aea/skills/base.py", line 440, in act_wrapper
    raise AEAActException(

aea.exceptions.AEAActException: An error occurred during act of behaviour valory/trader_abci:0.1.0/ToolSelectionBehaviour:
Traceback (most recent call last):

  File "/root/agent/vendor/valory/skills/decision_maker_abci/behaviours/tool_selection.py", line 81, in mech_tools
    raise ValueError("The mech's tools have not been set.")

ValueError: The mech's tools have not been set.

` in skills <bound method Behaviour.act_wrapper of <behaviours.TraderConsensusBehaviour object at 0x7f11f817a190>>! Please check logs.
Exception ignored in: <function StreamWriter.__del__ at 0x7f11fef0eb60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 395, in __del__
    self.close()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 343, in close
    return self._transport.close()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 860, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 761, in call_soon
    self._check_closed()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 519, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function StreamWriter.__del__ at 0x7f11fef0eb60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 395, in __del__
    self.close()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 343, in close
    return self._transport.close()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 860, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 761, in call_soon
    self._check_closed()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 519, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function StreamWriter.__del__ at 0x7f11fef0eb60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 395, in __del__
    self.close()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 343, in close
    return self._transport.close()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 860, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 761, in call_soon
    self._check_closed()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 519, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function StreamWriter.__del__ at 0x7f11fef0eb60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 395, in __del__
    self.close()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 343, in close
    return self._transport.close()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 860, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 761, in call_soon
    self._check_closed()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 519, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function StreamWriter.__del__ at 0x7f11fef0eb60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 395, in __del__
    self.close()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 343, in close
    return self._transport.close()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 860, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 761, in call_soon
    self._check_closed()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 519, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function StreamWriter.__del__ at 0x7f11fef0eb60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 395, in __del__
    self.close()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 343, in close
    return self._transport.close()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 860, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 761, in call_soon
    self._check_closed()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 519, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function StreamWriter.__del__ at 0x7f11fef0eb60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 395, in __del__
    self.close()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 343, in close
    return self._transport.close()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 860, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 761, in call_soon
    self._check_closed()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 519, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function StreamWriter.__del__ at 0x7f11fef0eb60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 395, in __del__
    self.close()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 343, in close
    return self._transport.close()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 860, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 761, in call_soon
    self._check_closed()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 519, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function StreamWriter.__del__ at 0x7f11fef0eb60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 395, in __del__
    self.close()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 343, in close
    return self._transport.close()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 860, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 761, in call_soon
    self._check_closed()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 519, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function StreamWriter.__del__ at 0x7f11fef0eb60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 395, in __del__
    self.close()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 343, in close
    return self._transport.close()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 860, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 761, in call_soon
    self._check_closed()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 519, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function StreamWriter.__del__ at 0x7f11fef0eb60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 395, in __del__
    self.close()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 343, in close
    return self._transport.close()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 860, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 761, in call_soon
    self._check_closed()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 519, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
Exception ignored in: <function StreamWriter.__del__ at 0x7f11fef0eb60>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 395, in __del__
    self.close()
  File "/usr/local/lib/python3.11/asyncio/streams.py", line 343, in close
    return self._transport.close()
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 860, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 761, in call_soon
    self._check_closed()
  File "/usr/local/lib/python3.11/asyncio/base_events.py", line 519, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
[2023-11-15 14:34:29,897][ERROR] Task was destroyed but it is pending!
task: <Task pending name='Task-155911' coro=<TcpServerChannel.receive_messages() running at /root/agent/vendor/valory/connections/abci/connection.py:953> wait_for=<Future pending cb=[Task.task_wakeup()]>>
[2023-11-15 14:34:29,897][ERROR] Task was destroyed but it is pending!
task: <Task pending name='Task-155913' coro=<TcpServerChannel.receive_messages() running at /root/agent/vendor/valory/connections/abci/connection.py:953> wait_for=<Future pending cb=[Task.task_wakeup()]>>
[2023-11-15 14:34:29,897][ERROR] Task was destroyed but it is pending!
task: <Task pending name='Task-155916' coro=<TcpServerChannel.receive_messages() running at /root/agent/vendor/valory/connections/abci/connection.py:953> wait_for=<Future pending cb=[Task.task_wakeup()]>>
[2023-11-15 14:34:29,897][ERROR] Task was destroyed but it is pending!
task: <Task pending name='Task-155919' coro=<TcpServerChannel.receive_messages() running at /root/agent/vendor/valory/connections/abci/connection.py:953> wait_for=<Future pending cb=[Task.task_wakeup()]>>
```

--- a/packages/packages.json
+++ b/packages/packages.json
@@ -1,13 +1,13 @@
 {
     "dev": {
         "skill/valory/market_manager_abci/0.1.0": "bafybeic7o4pclkhnugyn7js5g3asxuqhkxpvunlp3mpup7aovhg2fto22i",
-        "skill/valory/decision_maker_abci/0.1.0": "bafybeigdp4pgm4ons5pzptmclyglyxyqut266s4zutpmqj2e22wdpdkt5y",
-        "skill/valory/trader_abci/0.1.0": "bafybeiej7tgi23eihevyw6t4uvia63afwq7qyxosupvlpooi3hyeiot4vq",
+        "skill/valory/decision_maker_abci/0.1.0": "bafybeihnjjfzoiihujnjd2gzbvmayz7avo5myslvl24wvyqj7zq4ufkvw4",
+        "skill/valory/trader_abci/0.1.0": "bafybeihqqz7yvn33iirooeqk7e6iz4dr5grwrfbkavamis6bnfoz6yj5pu",
         "contract/valory/market_maker/0.1.0": "bafybeid5jnetbsusgxkpqzwyiqlwut55xktfbeloipxcemcuu5oopmqtl4",
-        "agent/valory/trader/0.1.0": "bafybeict4eqondm64jwst775rai7d2iw2m5x4rokjcoypmr3qv4vs6q5aa",
-        "service/valory/trader/0.1.0": "bafybeigwwsyu7keuvxz64ii3idce7ir4oqna7pnz4n6corrojd73kyatbq",
+        "agent/valory/trader/0.1.0": "bafybeigt6hd5cwuusdbu4u3ieqzvs4v4vihlr344rz55pb6neys55cwlre",
+        "service/valory/trader/0.1.0": "bafybeiaabwa7whlwpep5j2gpidnfvqgkgyo2vrw655t6epwtbh67umsb5m",
         "contract/valory/erc20/0.1.0": "bafybeiepg5ymxhtylgijs3mf3lezz6cwcsougbrv2gr4xcq3bp5wxusi64",
-        "skill/valory/tx_settlement_multiplexer_abci/0.1.0": "bafybeic6bveoijtjzvquphekmset7sondro4iknljf5qnylkadeeusiwp4",
+        "skill/valory/tx_settlement_multiplexer_abci/0.1.0": "bafybeiaqh4pujl2o2ginl5lbe7p7en7t2ncs25gohysxdks5kd4lmqav3q",
         "contract/valory/mech/0.1.0": "bafybeieo65rtidzgrx22la7z7azoqbisd35hnoqfw7l3xxctotvctjlkju",
         "contract/valory/realitio/0.1.0": "bafybeieoily22pasgkzzcyawhqyrsko7b52lghnvwuskhomavb7tlb6pxa",
         "contract/valory/realitio_proxy/0.1.0": "bafybeidx37xzjjmapwacedgzhum6grfzhp5vhouz4zu3pvpgdy5pgb2fr4",

--- a/packages/valory/agents/trader/aea-config.yaml
+++ b/packages/valory/agents/trader/aea-config.yaml
@@ -43,10 +43,10 @@ skills:
 - valory/reset_pause_abci:0.1.0:bafybeiencz2uytz6fqj5wg7mcunevxjh4xg5gd6pqcgkshgqywhcimweyi
 - valory/termination_abci:0.1.0:bafybeigainmpy37gb33ogbrpggjbkuasodhthh5oz4vbzjdfjozerpf224
 - valory/transaction_settlement_abci:0.1.0:bafybeigk3debp6dswutqsuls2lqfvyj4ghe6kwjc2zfinnsvj6hujynxtq
-- valory/tx_settlement_multiplexer_abci:0.1.0:bafybeic6bveoijtjzvquphekmset7sondro4iknljf5qnylkadeeusiwp4
+- valory/tx_settlement_multiplexer_abci:0.1.0:bafybeiaqh4pujl2o2ginl5lbe7p7en7t2ncs25gohysxdks5kd4lmqav3q
 - valory/market_manager_abci:0.1.0:bafybeic7o4pclkhnugyn7js5g3asxuqhkxpvunlp3mpup7aovhg2fto22i
-- valory/decision_maker_abci:0.1.0:bafybeigdp4pgm4ons5pzptmclyglyxyqut266s4zutpmqj2e22wdpdkt5y
-- valory/trader_abci:0.1.0:bafybeiej7tgi23eihevyw6t4uvia63afwq7qyxosupvlpooi3hyeiot4vq
+- valory/decision_maker_abci:0.1.0:bafybeihnjjfzoiihujnjd2gzbvmayz7avo5myslvl24wvyqj7zq4ufkvw4
+- valory/trader_abci:0.1.0:bafybeihqqz7yvn33iirooeqk7e6iz4dr5grwrfbkavamis6bnfoz6yj5pu
 - valory/staking_abci:0.1.0:bafybeia3acxumkjg566dvgagv3swz7htf5xpvwq6ovmecvp5j2zdxdoabe
 default_ledger: ethereum
 required_ledgers:

--- a/packages/valory/services/trader/service.yaml
+++ b/packages/valory/services/trader/service.yaml
@@ -7,7 +7,7 @@ license: Apache-2.0
 fingerprint:
   README.md: bafybeigtuothskwyvrhfosps2bu6suauycolj67dpuxqvnicdrdu7yhtvq
 fingerprint_ignore_patterns: []
-agent: valory/trader:0.1.0:bafybeict4eqondm64jwst775rai7d2iw2m5x4rokjcoypmr3qv4vs6q5aa
+agent: valory/trader:0.1.0:bafybeigt6hd5cwuusdbu4u3ieqzvs4v4vihlr344rz55pb6neys55cwlre
 number_of_agents: 4
 deployment: {}
 ---

--- a/packages/valory/skills/decision_maker_abci/behaviours/tool_selection.py
+++ b/packages/valory/skills/decision_maker_abci/behaviours/tool_selection.py
@@ -147,7 +147,8 @@ class ToolSelectionBehaviour(DecisionMakerBaseBehaviour):
             return True
 
         if res is None:
-            msg = f"Could not get the mech agent's tools from {self.mech_tools_api.url}."
+            url = self.mech_tools_api.url
+            msg = f"Could not get the mech agent's tools from {url}."
             self.context.logger.error(msg)
             self.mech_tools_api.increment_retries()
             return False

--- a/packages/valory/skills/decision_maker_abci/behaviours/tool_selection.py
+++ b/packages/valory/skills/decision_maker_abci/behaviours/tool_selection.py
@@ -143,6 +143,7 @@ class ToolSelectionBehaviour(DecisionMakerBaseBehaviour):
         if self.mech_tools_api.is_retries_exceeded():
             error = "Retries were exceeded while trying to get the mech agent's data."
             self.context.logger.error(error)
+            self.mech_tools_api.reset_retries()
             return True
 
         if res is None:

--- a/packages/valory/skills/decision_maker_abci/behaviours/tool_selection.py
+++ b/packages/valory/skills/decision_maker_abci/behaviours/tool_selection.py
@@ -283,6 +283,8 @@ class ToolSelectionBehaviour(DecisionMakerBaseBehaviour):
                 mech_tools = json.dumps(self.mech_tools)
                 policy = self.policy.serialize()
                 utilized_tools = json.dumps(self.utilized_tools, sort_keys=True)
+                self._store_policy()
+                self._store_available_mech_tools()
 
             payload = ToolSelectionPayload(
                 self.context.agent_address,
@@ -292,6 +294,4 @@ class ToolSelectionBehaviour(DecisionMakerBaseBehaviour):
                 selected_tool,
             )
 
-        self._store_policy()
-        self._store_available_mech_tools()
         yield from self.finish_behaviour(payload)

--- a/packages/valory/skills/decision_maker_abci/behaviours/tool_selection.py
+++ b/packages/valory/skills/decision_maker_abci/behaviours/tool_selection.py
@@ -147,7 +147,7 @@ class ToolSelectionBehaviour(DecisionMakerBaseBehaviour):
             return True
 
         if res is None:
-            msg = f"Could not get the mech agent's tools from {self.mech_tools_api.api_id}"
+            msg = f"Could not get the mech agent's tools from {self.mech_tools_api.url}."
             self.context.logger.error(msg)
             self.mech_tools_api.increment_retries()
             return False

--- a/packages/valory/skills/decision_maker_abci/behaviours/tool_selection.py
+++ b/packages/valory/skills/decision_maker_abci/behaviours/tool_selection.py
@@ -253,6 +253,9 @@ class ToolSelectionBehaviour(DecisionMakerBaseBehaviour):
     def _select_tool(self) -> Generator[None, None, Optional[int]]:
         """Select a Mech tool based on an e-greedy policy and return its index."""
         yield from self._get_tools()
+        if self._mech_tools is None:
+            return None
+
         self._set_policy()
         selected_idx = self.policy.select_tool()
         selected = self.mech_tools[selected_idx] if selected_idx is not None else "NaN"

--- a/packages/valory/skills/decision_maker_abci/skill.yaml
+++ b/packages/valory/skills/decision_maker_abci/skill.yaml
@@ -21,7 +21,7 @@ fingerprint:
   behaviours/reedem.py: bafybeiekq77eatvf7xczbnyd4s6met4ci57ethwwoyxf5mhrcq7vuwiaxu
   behaviours/round_behaviour.py: bafybeig4tdktyu6hapoqymnxh2bgpds547st6a44heue657wkctwe4gjvm
   behaviours/sampling.py: bafybeifzhm4sspdvt227ksl5hjn26offgqpwempgbcwbr6dq7gyi2a46sm
-  behaviours/tool_selection.py: bafybeigfr2frkljrxyfxs5p3j42equzehgaqtkyuxk6eiujyudr6ajqakm
+  behaviours/tool_selection.py: bafybeifpvrrbftlkjsoiysnasihzttousk6hxnukzrln6z2d33e3zexkje
   dialogues.py: bafybeigpwuzku3we7axmxeamg7vn656maww6emuztau5pg3ebsoquyfdqm
   fsm_specification.yaml: bafybeifnob3ceim2mj7lqagtnpwqjqqxs5eg3oiwc73gwm6x5i2dvvlcya
   handlers.py: bafybeihj33szgrcxnpd73s4nvluyxwwsvhjum2cuq3ilhhe6vfola3k7vy

--- a/packages/valory/skills/trader_abci/skill.yaml
+++ b/packages/valory/skills/trader_abci/skill.yaml
@@ -25,8 +25,8 @@ skills:
 - valory/transaction_settlement_abci:0.1.0:bafybeigk3debp6dswutqsuls2lqfvyj4ghe6kwjc2zfinnsvj6hujynxtq
 - valory/termination_abci:0.1.0:bafybeigainmpy37gb33ogbrpggjbkuasodhthh5oz4vbzjdfjozerpf224
 - valory/market_manager_abci:0.1.0:bafybeic7o4pclkhnugyn7js5g3asxuqhkxpvunlp3mpup7aovhg2fto22i
-- valory/decision_maker_abci:0.1.0:bafybeigdp4pgm4ons5pzptmclyglyxyqut266s4zutpmqj2e22wdpdkt5y
-- valory/tx_settlement_multiplexer_abci:0.1.0:bafybeic6bveoijtjzvquphekmset7sondro4iknljf5qnylkadeeusiwp4
+- valory/decision_maker_abci:0.1.0:bafybeihnjjfzoiihujnjd2gzbvmayz7avo5myslvl24wvyqj7zq4ufkvw4
+- valory/tx_settlement_multiplexer_abci:0.1.0:bafybeiaqh4pujl2o2ginl5lbe7p7en7t2ncs25gohysxdks5kd4lmqav3q
 - valory/staking_abci:0.1.0:bafybeia3acxumkjg566dvgagv3swz7htf5xpvwq6ovmecvp5j2zdxdoabe
 behaviours:
   main:

--- a/packages/valory/skills/tx_settlement_multiplexer_abci/skill.yaml
+++ b/packages/valory/skills/tx_settlement_multiplexer_abci/skill.yaml
@@ -21,7 +21,7 @@ protocols:
 - valory/ledger_api:1.0.0:bafybeige5agrztgzfevyglf7mb4o7pzfttmq4f6zi765y4g2zvftbyowru
 skills:
 - valory/abstract_round_abci:0.1.0:bafybeidau7loztcfy3mxvoqrv7otbpciemd2wf3lsxyjraq4dcvuvib25e
-- valory/decision_maker_abci:0.1.0:bafybeigdp4pgm4ons5pzptmclyglyxyqut266s4zutpmqj2e22wdpdkt5y
+- valory/decision_maker_abci:0.1.0:bafybeihnjjfzoiihujnjd2gzbvmayz7avo5myslvl24wvyqj7zq4ufkvw4
 - valory/staking_abci:0.1.0:bafybeia3acxumkjg566dvgagv3swz7htf5xpvwq6ovmecvp5j2zdxdoabe
 behaviours:
   main:

